odoo/tools/populate.py

```diff
@@ -1,180 +1,357 @@
+            END
+        """, field=SQL.identifier(field.name), postfix=postfix)
+
+
+@contextmanager
+def ignore_indexes(env, tablename):
+    """
+    Temporarily drop indexes on table to speed up insertion.
+    PKey and Unique indexes are kept for constraints
+    """
+    indexes = env.execute_query_dict(SQL("""
+        SELECT indexname AS name, indexdef AS definition
+          FROM pg_indexes
+         WHERE tablename = %s
+           AND indexname NOT LIKE %s
+           AND indexdef NOT LIKE %s
+    """, tablename, '%pkey', '%UNIQUE%'))
+    if indexes:
+        _logger.info('Dropping indexes on table %s...', tablename)
+        env.cr.execute(SQL(';').join(
+            SQL('DROP INDEX %s CASCADE', SQL.identifier(index['name']))
+            for index in indexes
+        ))
+        yield
+        _logger.info('Adding indexes back on table %s...', tablename)
+        env.cr.execute(';'.join(index['definition'] for index in indexes))
+    else:
+        yield
+
+
+@contextmanager
+def ignore_fkey_constraints(env):
+    """
+    Disable Fkey constraints checks by setting the session to replica.
+    """
+    env.cr.execute('SET session_replication_role TO replica')
+    yield
+    env.cr.execute('RESET session_replication_role')
+
+
+def field_need_variation(env, model, field):
+    """
+    Return True/False depending on if the field needs to be varied.
+    Might be necessary in the case of:
+    - unique constraints
+    - varying dates for better distribution
+    - field will be part of _rec_name_search, therefor variety is needed for effective searches
+    - field has a trigram index on it
+    """
+    def is_unique(_env, _model, _field):
+        """
+        An unique constraint is enforced by Postgres as an unique index,
+        whether it's defined as a constraint on the table, or as an manual unique index.
+        Both type of constraint are present in the index catalog
+        """
+        query = SQL("""
+        SELECT EXISTS(SELECT 1
+              FROM pg_index idx
+                   JOIN pg_class t ON t.oid = idx.indrelid
+                   JOIN pg_class i ON i.oid = idx.indexrelid
+                   JOIN pg_attribute a ON a.attnum = ANY (idx.indkey) AND a.attrelid = t.oid
+              WHERE t.relname = %s  -- tablename
+                AND a.attname = %s  -- column
+                AND idx.indisunique = TRUE) AS is_unique;
+        """, _model._table, _field.name)
+        return _env.execute_query(query)[0][0]
+
+    # many2one fields are not considered, as a name_search would resolve it to the _rec_names_search of the related model
+    if model._rec_names_search and field.name in model._rec_names_search and field.type != 'many2one':
+        return True
+    if field.type in ('date', 'datetime'):
+        return True
+    if field.index == 'trigram':
+        return True
+    return is_unique(env, model, field)
+
+
+def variate_field(env, model, field, table_alias, series_alias, factors):
```

Comment:
Again, if `model` is a recordset, `env` is redundant.
```suggestion
def get_field_variation(env, model, field, table_alias, series_alias, factors):
```
And make the return type simpler (always `SQL`).